gtk4/src/subclass/widget.rs

```diff
@@ -123,3 +126,126 @@ unsafe extern "C" fn widget_direction_changed<T: ObjectSubclass>(
 
     imp.direction_changed(&wrap, dirwrap)
 }
+
+pub trait WidgetSubclass
```

Comment:
The naming scheme is different :) The way how such things are named elsewhere would be `WidgetClassSubclassExt`

Reply:
And this needs a blanket impl `unsafe impl WidgetClassSubclassExt for glib::Class<Widget> {}`

(And the crate itself should be unsafe, the functions probably not)

Reply:
> And this needs a blanket impl unsafe impl WidgetClassSubclassExt for glib::Class<Widget> {}

These functions should only be used in the context of subclassing during `class_init()`; Would that blanket impl apply it to every widget?

Reply:
It doesn't work with the trait bounds ` ObjectSubclass + WidgetImpl`

Reply:
Which in turn means we can't use `Self::Class` in `bind_template_children()`

Reply:
Check gstreamer-rs for various patterns for class methods that should only be used for subclassing :)

Reply:
I've made a change so that it's `unsafe impl<T: ObjectSubclass + WidgetImpl> WidgetClassSubclassExt for T {}`. Is that okay?

Reply:
`unsafe impl WidgetClassSubclassExt for glib::Class<gtk::Widget> {}` did not work?

Another variant that I have is `unsafe impl<T: ClassStruct> WidgetClassSubclassExt for T where     T::Type: WidgetImpl`.

And then the variant I added in the new comment :)

But I think in the end these are all equivalent in practice. But it would be good to have an example of this, and especially one that makes sure this actually doesn't only work for direct subclasses of `gtk::Widget` but also e.g. `gtk::Box`.

Reply:
And it would be good to have it implemented on the class struct in some way, so that inside `class_init()` you can just call `class.blabla()`.

Anyway, if you provide an example in the repo I can take a look at such things. Otherwise I can just give suggestions that might not work :)

Reply:
Added an example.

Reply:
Something like this would work for example. Then you can call functions directly on the class struct. Only ugly part left then is `Self::bind_template_children(klass)` but I'm not entirely sure how to solve that.

```diff
diff --git a/examples/src/bin/composite_template.rs b/examples/src/bin/composite_template.rs
index de22d7f..6bec20f 100644
--- a/examples/src/bin/composite_template.rs
+++ b/examples/src/bin/composite_template.rs
@@ -52,7 +52,7 @@ impl ObjectSubclass for ExApplicationWindowPriv {
     // at once.
     fn class_init(klass: &mut Self::Class) {
         let template = include_bytes!("composite_template.ui");
-        Self::set_template(klass, template);
+        klass.set_template(template);
         Self::bind_template_children(klass);
     }
 }
diff --git a/gtk4-macros/src/composite_template_derive.rs b/gtk4-macros/src/composite_template_derive.rs
index 9b95527..45a3365 100644
--- a/gtk4-macros/src/composite_template_derive.rs
+++ b/gtk4-macros/src/composite_template_derive.rs
@@ -28,8 +28,7 @@ fn gen_template_child_bindings(fields: &syn::Fields) -> TokenStream {
             }
 
             quote! {
-                Self::bind_template_child_with_offset(
-                    klass,
+                klass.bind_template_child_with_offset(
                     &#value_id,
                     #crate_ident::offset_of!(Self => #ident),
                 );
diff --git a/gtk4/src/subclass/widget.rs b/gtk4/src/subclass/widget.rs
index 965b335..962dbf2 100644
--- a/gtk4/src/subclass/widget.rs
+++ b/gtk4/src/subclass/widget.rs
@@ -108,10 +108,10 @@ unsafe extern "C" fn widget_direction_changed<T: WidgetImpl>(
     imp.direction_changed(wrap.unsafe_cast_ref(), dirwrap)
 }
 
-pub unsafe trait WidgetClassSubclassExt: ObjectSubclass + WidgetImpl {
-    fn set_template_bytes(klass: &mut <Self as ObjectSubclass>::Class, template: &glib::Bytes) {
+pub unsafe trait WidgetClassSubclassExt: ClassStruct {
+    fn set_template_bytes(&mut self, template: &glib::Bytes) {
         unsafe {
-            let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+            let type_class = self as *mut _ as *mut gobject_sys::GTypeClass;
             let widget_class =
                 gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
                     as *mut gtk_sys::GtkWidgetClass;
@@ -119,22 +119,19 @@ pub unsafe trait WidgetClassSubclassExt: ObjectSubclass + WidgetImpl {
         }
     }
 
-    fn set_template(klass: &mut <Self as ObjectSubclass>::Class, template: &[u8]) {
+    fn set_template(&mut self, template: &[u8]) {
         let template_bytes = glib::Bytes::from(template);
-        Self::set_template_bytes(klass, &template_bytes);
+        self.set_template_bytes(&template_bytes);
     }
 
-    fn set_template_static(klass: &mut <Self as ObjectSubclass>::Class, template: &'static [u8]) {
+    fn set_template_static(&mut self, template: &'static [u8]) {
         let template_bytes = glib::Bytes::from_static(template);
-        Self::set_template_bytes(klass, &template_bytes);
+        self.set_template_bytes(&template_bytes);
     }
 
-    fn set_template_from_resource(
-        klass: &mut <Self as ObjectSubclass>::Class,
-        resource_name: &str,
-    ) {
+    fn set_template_from_resource(&mut self, resource_name: &str) {
         unsafe {
-            let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+            let type_class = self as *mut _ as *mut gobject_sys::GTypeClass;
             let widget_class =
                 gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
                     as *mut gtk_sys::GtkWidgetClass;
@@ -145,9 +142,9 @@ pub unsafe trait WidgetClassSubclassExt: ObjectSubclass + WidgetImpl {
         }
     }
 
-    fn bind_template_child(klass: &mut <Self as ObjectSubclass>::Class, name: &str) {
+    fn bind_template_child(&mut self, name: &str) {
         unsafe {
-            let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+            let type_class = self as *mut _ as *mut gobject_sys::GTypeClass;
             let widget_class =
                 gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
                     as *mut gtk_sys::GtkWidgetClass;
@@ -162,17 +159,19 @@ pub unsafe trait WidgetClassSubclassExt: ObjectSubclass + WidgetImpl {
 
     #[allow(clippy::missing_safety_doc)]
     unsafe fn bind_template_child_with_offset<T>(
-        klass: &mut <Self as ObjectSubclass>::Class,
+        &mut self,
         name: &str,
-        offset: field_offset::FieldOffset<Self, TemplateChild<T>>,
+        offset: field_offset::FieldOffset<Self::Type, TemplateChild<T>>,
     ) where
         T: ObjectType + FromGlibPtrNone<*mut <T as ObjectType>::GlibType>,
     {
-        let type_class = klass as *mut _ as *mut gobject_sys::GTypeClass;
+        let type_class = self as *mut _ as *mut gobject_sys::GTypeClass;
         let widget_class =
             gobject_sys::g_type_check_class_cast(type_class, gtk_sys::gtk_widget_get_type())
                 as *mut gtk_sys::GtkWidgetClass;
-        let private_offset = Self::type_data().as_ref().private_offset;
+        let private_offset = <Self::Type as ObjectSubclass>::type_data()
+            .as_ref()
+            .private_offset;
         gtk_sys::gtk_widget_class_bind_template_child_full(
             widget_class,
             name.to_glib_none().0,
@@ -182,7 +181,7 @@ pub unsafe trait WidgetClassSubclassExt: ObjectSubclass + WidgetImpl {
     }
 }
 
-unsafe impl<T: ObjectSubclass + WidgetImpl> WidgetClassSubclassExt for T {}
+unsafe impl<T: ClassStruct> WidgetClassSubclassExt for T where T::Type: WidgetImpl {}
 
 #[derive(Debug, PartialEq, Eq)]
 #[repr(transparent)]
@@ -217,6 +216,6 @@ where
     }
 }
 
-pub trait CompositeTemplate: WidgetClassSubclassExt {
+pub trait CompositeTemplate: WidgetImpl {
     fn bind_template_children(klass: &mut Self::Class);
 }
```